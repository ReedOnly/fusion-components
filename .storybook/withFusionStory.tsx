import React from 'react';
import Header from '../src/components/core/Header';
import Content from '../src/components/core/Content';
import Overlay from '../src/components/general/ApplicationGuidance/components/Overlay';

export default (title: string, padding: number = 16) => (stories: () => React.ReactNode) => {
    return (
        <>
            <Overlay fixed>
                <Header aside={null} start={null} content={null} />
            </Overlay>
            <Content>
                <Overlay>
                    <div
                        style={{
                            padding,
                            minHeight: '100%',
                            width: '100%',
                            maxWidth: '100%',
                            boxSizing: 'border-box',
                        }}
                    >
                        {stories()}
                    </div>
                </Overlay>
            </Content>
        </>
    );
};
