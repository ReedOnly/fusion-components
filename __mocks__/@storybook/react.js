import renderer from 'react-test-renderer';
import addons, { mockChannel } from '@storybook/addons';
import { withFusionContext } from '../../.storybook/withFusionContext';

addons.setChannel(mockChannel());

export const action = (actionName) => jest.fn();
const fusionContext = withFusionContext();
export const storiesOf = (groupName) => {
    const api = {
        decorators: [],
        parameters: {},

        add(storyName, story) {
            if (api.parameters.skipTest) {
                return;
            }

            describe(groupName, () => {
                it(storyName, () => {
                    const component = renderer.create(fusionContext(story));

                    // @TODO Enable custom names once released:
                    // > https://github.com/facebook/jest/pull/2094
                    expect(component.toJSON())
                        .toMatchSnapshot
                        // `${groupName}.${storyName}`
                        ();
                });
            });

            return api;
        },

        addDecorator() {
            return api;
        },

        addParameters(parameters) {
            api.parameters = { ...api.parameters, ...parameters };
            return api;
        },
    };

    return api;
};

export const addDecorator = () => {};
