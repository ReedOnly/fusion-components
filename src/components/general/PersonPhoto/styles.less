
.photoContainer {
    display: flex;
    border-radius: 50%;
    background-size: cover;
    background-position: center;
    background-repeat: no-repeat;
    position: absolute;

    &.xlarge {
        &.comfortable {
            // 56x56
            width: calc(var(--grid-unit) * 7px);
            height: calc(var(--grid-unit) * 7px);
        }

        &.compact {
            // 48x48
            width: calc(var(--grid-unit) * 6px);
            height: calc(var(--grid-unit) * 6px);
        }
    }

    &.large {
        &.comfortable {
            // 40x40
            width: calc(var(--grid-unit) * 5px);
            height: calc(var(--grid-unit) * 5px);
        }

        &.compact {
            // 32x32
            width: calc(var(--grid-unit) * 4px);
            height: calc(var(--grid-unit) * 4px);
        }
    }

    &.medium {
        &.comfortable {
            // 32x32
            width: calc(var(--grid-unit) * 4px);
            height: calc(var(--grid-unit) * 4px);
        }

        &.compact {
            // 24x24
            width: calc(var(--grid-unit) * 3px);
            height: calc(var(--grid-unit) * 3px);
        }
    }

    &.small {
        &.comfortable {
            // 24x24
            width: calc(var(--grid-unit) * 3px);
            height: calc(var(--grid-unit) * 3px);
        }

        &.compact {
            // 16x16
            width: calc(var(--grid-unit) * 2px);
            height: calc(var(--grid-unit) * 2px);
        }
    }

    .affiliationContainer {
        border-radius: 50%;
        position: absolute;
        bottom: 0;
        right: 0;

        &.affiliate {
            background: var(--color-secondary-alt2);
        }
    
        &.consultant, &.externalHire {
            background: var(--color-red);
        }
    
        &.xlarge {
            &.comfortable {
                // 24x24
                width: calc(var(--grid-unit) * 3px);
                height: calc(var(--grid-unit) * 3px);
            }
    
            &.compact {
                // 16x16
                width: calc(var(--grid-unit) * 2px);
                height: calc(var(--grid-unit) * 2px);
            }
        }
    
        &.large, &.medium {
            &.comfortable {
                // 16x16
                width: calc(var(--grid-unit) * 2px);
                height: calc(var(--grid-unit) * 2px);
            }
    
            &.compact {
                // 8x8
                width: calc(var(--grid-unit) * 1px);
                height: calc(var(--grid-unit) * 1px);
            }
        }
    
        &.small {
            &.comfortable {
                // 12x12
                width: calc(var(--grid-unit) * 1.5px);
                height: calc(var(--grid-unit) * 1.5px);
            }
    
            &.compact {
                // 8x8
                width: calc(var(--grid-unit) * 1px);
                height: calc(var(--grid-unit) * 1px);
            }
        }
    
    }
}
