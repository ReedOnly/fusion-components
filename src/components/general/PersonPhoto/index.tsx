import React from 'react';
import styles from './styles.less';
import classNames from 'classnames';
import { useFusionContext, useComponentDisplayClassNames } from '@equinor/fusion';
import ConsultantIcon from './icons/ConsultantIcon';
import ExternalHireIcon from './icons/ExternalHireIcon';
import AffiliateIcon from './icons/AffiliateIcon';

export type PhotoSize = 'xlarge' | 'large' | 'medium' | 'small';
export type Affiliation = 'externalhire' | 'consultant' | 'affiliate';

export type PersonPhotoProps = {
    personId: string;
    affiliation: Affiliation;
    size: PhotoSize;
    title?: string;
};

const friendlyAffiliationName = (affiliation: string) => {
    switch (affiliation) {
        case 'externalhire':
            return 'External Hire';
        case 'consultant':
            return 'Consultant';
        case 'affiliate':
            return 'Affiliate';
    }
};

const iconSizes = {
    xlarge: { width: 24, height: 24 },
    large: { width: 16, height: 16 },
    medium: { width: 16, height: 16 },
    small: { width: 12, height: 12 },
};

export default ({ personId, affiliation, size, title }: PersonPhotoProps) => {
    const fusionContext = useFusionContext();
    const urlToPhoto = fusionContext.http.resourceCollections.people.getPersonPhoto(personId);

    const photoClassNames = classNames(
        styles.photoContainer,
        useComponentDisplayClassNames(styles),
        {
            [styles.xlarge]: size === 'xlarge',
            [styles.large]: size === 'large',
            [styles.medium]: size === 'medium',
            [styles.small]: size === 'small',
        }
    );

    const iconClassNames = classNames(
        styles.affiliationContainer,
        useComponentDisplayClassNames(styles),
        {
            [styles.xlarge]: size === 'xlarge',
            [styles.large]: size === 'large',
            [styles.medium]: size === 'medium',
            [styles.small]: size === 'small',
            [styles.externalHire]: affiliation === 'externalhire',
            [styles.consultant]: affiliation === 'consultant',
            [styles.affiliate]: affiliation === 'affiliate',
        }
    );

    return (
        <React.Fragment>
            <img className={photoClassNames} src="https://via.placeholder.com/150" />
            <div
                title={title}
                className={photoClassNames}
                style={{
                    backgroundImage: `url('${urlToPhoto}')`,
                }}
            >
                <div className={iconClassNames} title={friendlyAffiliationName(affiliation)}>
                    {affiliation === 'consultant' && (
                        <ConsultantIcon
                            width={iconSizes[size].width}
                            height={iconSizes[size].height}
                        />
                    )}
                    {affiliation === 'externalhire' && (
                        <ExternalHireIcon
                            width={iconSizes[size].width}
                            height={iconSizes[size].height}
                        />
                    )}
                    {affiliation === 'affiliate' && (
                        <AffiliateIcon
                            width={iconSizes[size].width}
                            height={iconSizes[size].height}
                        />
                    )}
                </div>
            </div>
        </React.Fragment>
    );
};
