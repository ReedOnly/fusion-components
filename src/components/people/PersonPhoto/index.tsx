import React, { useState, useEffect, useMemo } from 'react';
import styles from './styles.less';
import classNames from 'classnames';
import {
    useFusionContext,
    useComponentDisplayClassNames,
    usePersonDetails,
    PersonDetails,
} from '@equinor/fusion';

import { useTooltipRef } from '@equinor/fusion-components';
import FallbackImage from './FallbackImage';
import AccountTypeBadge from './AccountTypeBadge';

export type PhotoSize = 'xlarge' | 'large' | 'medium' | 'small';

export type PersonPhotoProps = {
    personId?: string;
    person?: PersonDetails;
    size?: PhotoSize;
    hideTooltip?: boolean;
};

const getDefaultPerson = (): PersonDetails => ({
    azureUniqueId: 'string',
    name: 'No name',
    mail: 'noname@equinor.com',
    jobTitle: 'www',
    department: 'string',
    mobilePhone: 'string',
    officeLocation: 'string',
    upn: 'string',
    accountType: 'Employee',
    company: { id: 'id', name: 'name' },
});

export default ({ personId, person, hideTooltip, size = 'medium' }: PersonPhotoProps) => {
    const fusionContext = useFusionContext();
    const [isFallbackImage, setIsFallbackImage] = useState(false);
    const [currentPerson, setCurrentPerson] = useState<PersonDetails>(getDefaultPerson());

    const urlToPhoto = useMemo(() => {
        const peopleUrls = fusionContext.http.resourceCollections.people;

        if (person) {
            return peopleUrls.getPersonPhoto(person.azureUniqueId);
        } else if (personId) {
            return peopleUrls.getPersonPhoto(personId);
        }
    }, [personId, person, fusionContext]);

    const [error, _, personDetails] = personId ? usePersonDetails(personId) : [null, false, person];

    useEffect(() => {
        if (!error && personDetails) {
            setCurrentPerson(personDetails);
        } else if (error) {
            setCurrentPerson(getDefaultPerson());
        }
    }, [error, personDetails]);

    const photoClassNames = classNames(
        styles.photoContainer,
        useComponentDisplayClassNames(styles),
        {
            [styles.xlarge]: size === 'xlarge',
            [styles.large]: size === 'large',
            [styles.medium]: size === 'medium',
            [styles.small]: size === 'small',
        }
    );

    useEffect(() => {
        setIsFallbackImage(false);
        
        const image = new Image();
        image.onerror = () => setIsFallbackImage(true);
        image.src = `${urlToPhoto}`;
    }, [urlToPhoto]);

    const imageStyle = !isFallbackImage ? { backgroundImage: `url(${urlToPhoto})` } : {};
    const nameTooltipRef = useTooltipRef(hideTooltip ? '' : currentPerson.name);

    return (
        <div ref={nameTooltipRef} className={photoClassNames} style={imageStyle}>
            {isFallbackImage && <FallbackImage size={size} />}
            <AccountTypeBadge
                currentPerson={personDetails || getDefaultPerson()}
                size={size}
                hideTooltip={hideTooltip}
            />
        </div>
    );
};
