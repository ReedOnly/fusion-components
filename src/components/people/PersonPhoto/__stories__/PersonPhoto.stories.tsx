import React from 'react';
import { storiesOf } from '@storybook/react';
import withFusionStory from '../../../../../.storybook/withFusionStory';
import PersonPhoto from '../index';
import { withKnobs, number, color, select } from '@storybook/addon-knobs';

const stories = storiesOf('People|PersonPhoto', module);
stories.addDecorator(withKnobs);
stories.addDecorator(withFusionStory('PersonPhoto'));

const personIds = {
    Consultant: 'e92c631b-94ae-4670-8f1e-60cdc2122edc', // Morten Salte
    Employee: 'e2d6d1a4-5b48-4a1d-8db1-08a5043dc658', // Egil Sagstad
    External: 'e1',
    Invalid: 'invalid',
};

const sizes = {
    XLarge: 'xlarge',
    Large: 'large',
    Medium: 'medium',
    Small: 'small',
};

const PersonPhotoStory = () => {
    const personId = select('Person type', personIds, 'e');
    const size = select('Size', sizes, 'xlarge');

    return <PersonPhoto personId={personId} size={size} />;
};

stories.add('Default', () => <PersonPhotoStory />);
