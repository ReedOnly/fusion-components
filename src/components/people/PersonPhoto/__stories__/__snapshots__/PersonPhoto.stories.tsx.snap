// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`People|PersonPhoto Default 1`] = `
<div
  className="container comfortable"
>
  <div
    className="container comfortable"
  >
    <div
      className="photoContainer comfortable xlarge"
      style={
        Object {
          "backgroundImage": "url()",
        }
      }
    >
      <div>
        <div
          className="popoverDetails hidePopover"
        />
      </div>
    </div>
  </div>
  <div
    className="overlay"
  />
  <div
    className="snacks"
  />
  <div
    className="dialog"
  />
</div>
`;
